universal-idl-program/program/src/instructions/create_idl/mod.rs

```diff
@@ -0,0 +1,208 @@
+            })
+        } else {
+            Err(assert_with_msg(
+                false,
+                ProgramError::IllegalOwner,
+                lazy_format!(
+                    "Program account must be owned by upgradeable or frozen loader, instead of: {}",
+                    program.owner
+                ),
+            )
+            .unwrap_err())
+        }
+    }
+}
+
+pub struct CreateIdlCtx<'a, 'info> {
+    pub payer: &'a AccountInfo<'info>,
+    pub program_authority: &'a AccountInfo<'info>,
+    pub idl_account: &'a AccountInfo<'info>,
+    pub authority_source: AuthoritySource<'a, 'info>,
+    pub system_program: &'a AccountInfo<'info>,
+}
+
+impl<'a, 'info> CreateIdlCtx<'a, 'info> {
+    pub fn load(accounts: &'a [AccountInfo<'info>]) -> Result<Self, ProgramError> {
+        let accounts_iter = &mut accounts.iter();
+
+        // payer
+        let payer = next_account_info(accounts_iter)?;
+        assert_signer(payer, "payer")?;
+        assert_mut(payer, "payer")?;
+
+        // program authority
+        let program_authority = next_account_info(accounts_iter)?;
+        assert_signer(program_authority, "program_authority")?;
+
+        // idl_account
+        let idl_account = next_account_info(accounts_iter)?;
+        assert_mut(idl_account, "idl_account")?;
+
+        // authority_source
+        let authority_source = AuthoritySource::load(accounts_iter)?;
+        authority_source.check_authority(program_authority.key)?;
+
+        // system_program
+        let system_program = next_account_info(accounts_iter)?;
+        assert_system_program(system_program)?;
+
+        Ok(Self {
+            payer,
+            program_authority,
+            idl_account,
+            authority_source,
+            system_program,
+        })
+    }
+
+    pub fn get_program(&self) -> Pubkey {
+        match self.authority_source {
+            AuthoritySource::UpgradeableAuthority {
+                upgradeable_program,
+                ..
+            } => upgradeable_program.key.clone(),
+            AuthoritySource::FrozenProgramAuthority { frozen_program, .. } => {
+                frozen_program.key.clone()
+            }
+        }
+    }
+}
+
+pub fn handler(ctx: CreateIdlCtx) -> ProgramResult {
+    let idl_seeds = assert_idl_seeds(&ctx.get_program(), ctx.idl_account.key)?;
+    let data_len = std::mem::size_of::<Idl>();
+    invoke_signed(
+        &create_account(
+            ctx.payer.key,
+            ctx.idl_account.key,
+            Rent::get()?.minimum_balance(data_len as usize),
+            data_len as u64,
+            &id(),
+        ),
+        &[ctx.payer.clone(), ctx.idl_account.clone()],
+        &[&idl_seeds
+            .iter()
+            .map(|s| s.as_slice())
+            .collect::<Vec<&[u8]>>()],
+    )?;
+
+    let mut idl: Idl = Idl::new();
+    idl.authority = *ctx.program_authority.key;
+    idl.slot = Clock::get()?.slot;
+    idl.save(ctx.idl_account)?;
```

Comment:
I think this is potentially suspect due to the memory limits (depending on how borsh deals with Vec<u8>). As long as there is not an intermediate serialized representation on the stack I think this call is safe, otherwise you might blow through the 4KB limit